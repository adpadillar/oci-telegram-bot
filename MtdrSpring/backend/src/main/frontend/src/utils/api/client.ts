import { z } from "zod";

export const userResponseValidator = z.object({
  id: z.number(),
  title: z.string().nullable(),
  role: z.enum(["developer", "manager", "user-pending-activation"]),
  firstName: z.string(),
  lastName: z.string(),
});

export const userRequestValidator = z.object({
  telegramId: z.string().nullable(),
  firstName: z.string(),
  lastName: z.string(),
  role: z.enum(["developer", "manager", "user-pending-activation"]),
  title: z.string().nullable(),
});

export type UserResponse = z.infer<typeof userResponseValidator>;
export type UserRequest = z.infer<typeof userRequestValidator>;

export const sprintValidator = z.object({
  id: z.number(),
  name: z.string(),
});

export const projectValidator = z.object({
  id: z.number(),
  name: z.string(),
});

export const taskResponseValidator = z.object({
  id: z.number(),
  description: z.string(),
  createdAt: z.string().transform((value) => new Date(value)),
  status: z.string(),
  createdBy: userResponseValidator,
  assignedTo: userResponseValidator.nullable(),
  estimateHours: z.number().nullable(),
  realHours: z.number().nullable(),
  sprint: sprintValidator.nullable(),
  project: projectValidator,
  category: z.string().nullable(),
});

export const taskRequestValidator = z.object({
  description: z.string(),
  status: z.enum(["created", "in-progress", "in-review", "testing", "done"]),
  createdBy: z.number(),
  assignedTo: z.number().nullable(),
  estimateHours: z.number().nullable(),
  realHours: z.number().nullable(),
  sprint: z.number().nullable(),
  category: z.enum(["feature", "bug", "issue"]).nullable(),
});

export type TaskResponse = z.infer<typeof taskResponseValidator>;
export type TaskRequest = z.infer<typeof taskRequestValidator>;

function createApiClient(baseUrl: string) {
  // for now we are hardcoding the project ID to 1
  const urlWithProject = `${baseUrl}/1`;

  async function getProjectManager() {
    const response = await fetch(`${urlWithProject}/users`);
    const responseJson = await response.json();

    const safeParsed = userResponseValidator.array().safeParse(responseJson);

    if (!safeParsed.success) {
      console.error("Error", safeParsed.error);
      throw new Error("Invalid data");
    }

    console.log("Safe parsed", safeParsed.data);

    const manager = safeParsed.data.find((user) => user.role === "manager");

    if (!manager) {
      throw new Error("No manager found");
    }

    return manager;
  }

  async function getProjectDevelopers() {
    const response = await fetch(`${urlWithProject}/users`);
    const responseJson = await response.json();

    const safeParsed = userResponseValidator.array().safeParse(responseJson);

    if (!safeParsed.success) {
      console.error("Error", safeParsed.error);
      throw new Error("Invalid data");
    }

    console.log("Safe parsed", safeParsed.data);

    return safeParsed.data.filter(
      (user) =>
        user.role === "developer" || user.role === "user-pending-activation"
    );
  }

  async function listTasks() {
    const response = await fetch(`${urlWithProject}/tasks`);
    const tasks = await response.json();

    const safeParsed = taskResponseValidator.array().safeParse(tasks);

    if (!safeParsed.success) {
      console.error("Error", safeParsed.error);
      throw new Error("Invalid data");
    }

    console.log("Safe parsed", safeParsed.data);

    return safeParsed.data.sort(
      (a, b) => b.createdAt.getTime() - a.createdAt.getTime()
    );
  }

  async function createTask(task: Omit<TaskRequest, "createdBy">) {
    const manager = await getProjectManager();

    return fetch(`${urlWithProject}/tasks`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        ...task,
        createdBy: manager.id, // updated to use the manager's ID
      }),
    });
  }

  async function patchTask(id: number, task: Partial<TaskRequest>) {
    return fetch(`${urlWithProject}/tasks/${id}`, {
      method: "PATCH",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(task),
    });
  }

  async function deleteTask(id: number) {
    return fetch(`${urlWithProject}/tasks/${id}`, {
      method: "DELETE",
    });
  }

  async function patchUser(id: number, user: Partial<UserRequest>) {
    return fetch(`${urlWithProject}/users/${id}`, {
      method: "PATCH",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(user),
    });
  }

  async function updateUserStatus(id: number, status: "developer" | "manager") {
    return patchUser(id, { role: status });
  }

  return {
    tasks: {
      list: listTasks,
      create: createTask,
      patch: patchTask,
      delete: deleteTask,
    },
    users: {
      getManager: getProjectManager,
      getDevelopers: getProjectDevelopers,
      patch: patchUser,
      updateStatus: updateUserStatus,
    },
  };
}

export const api = createApiClient("http://localhost:8081/api");
