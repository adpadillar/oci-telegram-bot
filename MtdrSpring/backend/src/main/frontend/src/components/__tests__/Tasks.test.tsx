import { render, screen, waitFor, fireEvent } from "@testing-library/react";
import { describe, it, vi, expect } from "vitest";
import "@testing-library/jest-dom";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { MemoryRouter } from "react-router-dom";
import Tasks from "../Tasks";
import * as client from "../../utils/api/client";

// npx vitest run
// Mock the API client
vi.mock("../../utils/api/client", async () => {
  const original = await vi.importActual<typeof client>(
    "../../utils/api/client",
  );

  const mockUsers = [
    { id: 1, firstName: "Ana", lastName: "Ramírez", role: "developer" },
    { id: 2, firstName: "Luis", lastName: "Hernández", role: "developer" },
  ];

  const mockTasks = [
    {
      id: 1,
      description: "Fix login bug",
      status: "done",
      estimateHours: 4,
      realHours: 5,
      createdById: 1,
      assignedToId: 1,
      sprintId: 1,
      projectId: 1,
      createdAt: new Date().toISOString(),
      category: "bug",
    },
    {
      id: 2,
      description: "Implement UI",
      status: "in-progress",
      estimateHours: 6,
      realHours: null,
      createdById: 2,
      assignedToId: 2,
      sprintId: 1,
      projectId: 1,
      createdAt: new Date().toISOString(),
      category: "feature",
    },
  ];

  const mockSprints = [
    {
      id: 1,
      name: "Sprint 1",
      description: "Initial sprint",
      startedAt: new Date("2024-03-01").toISOString(),
      endsAt: new Date("2024-03-07").toISOString(),
      projectId: 1,
    },
  ];

  return {
    ...original,
    api: {
      tasks: {
        list: vi.fn().mockResolvedValue(mockTasks),
        create: vi.fn().mockResolvedValue({ ...mockTasks[0], id: 3 }),
        patch: vi
          .fn()
          .mockResolvedValue({ ...mockTasks[0], description: "Updated task" }),
        delete: vi.fn().mockResolvedValue({}),
      },
      users: {
        getUsers: vi.fn().mockResolvedValue(mockUsers),
      },
      sprints: {
        getSprints: vi.fn().mockResolvedValue(mockSprints),
      },
    },
  };
});

// Render helper
function renderTasks() {
  const queryClient = new QueryClient();
  return render(
    <MemoryRouter>
      <QueryClientProvider client={queryClient}>
        <Tasks />
      </QueryClientProvider>
    </MemoryRouter>,
  );
}

describe("Tasks Component", () => {
  it("renders the Tasks title", async () => {
    renderTasks();
    await waitFor(() => {
      expect(screen.getByText("Tasks")).toBeInTheDocument();
    });
  });

  it("displays the correct number of tasks", async () => {
    renderTasks();
    await waitFor(() => {
      expect(screen.getByText("2 tasks")).toBeInTheDocument();
    });
  });

  it("renders task cards with correct information", async () => {
    renderTasks();
    await waitFor(() => {
      expect(screen.getByText("Fix login bug")).toBeInTheDocument();
      expect(screen.getByText("Implement UI")).toBeInTheDocument();
    });
  });

  it("searches tasks by description", async () => {
    renderTasks();
    await waitFor(() => {
      const searchInput = screen.getByPlaceholderText("Search tasks...");
      fireEvent.change(searchInput, { target: { value: "login" } });

      expect(screen.getByText("Fix login bug")).toBeInTheDocument();
      expect(screen.queryByText("Implement UI")).not.toBeInTheDocument();
    });
  });

  it("displays task status correctly", async () => {
    renderTasks();
    await waitFor(() => {
      expect(screen.getByText("Done")).toBeInTheDocument();
      expect(screen.getByText("In Progress")).toBeInTheDocument();
    });
  });

  it("shows task estimates and actual hours", async () => {
    renderTasks();
    await waitFor(() => {
      expect(screen.getByText("4h")).toBeInTheDocument();
      expect(screen.getByText("5h")).toBeInTheDocument();
      expect(screen.getByText("6h")).toBeInTheDocument();
    });
  });

  it("allows editing task description (state change)", async () => {
    renderTasks();
    // Wait for edit buttons to appear
    await waitFor(() => {
      expect(screen.getAllByTitle("Edit task")[0]).toBeInTheDocument();
    });
    // Open edit modal for first task
    fireEvent.click(screen.getAllByTitle("Edit task")[0]);
    // Change description
    const descriptionInput = screen.getByLabelText("Task Description") as HTMLInputElement;
    fireEvent.change(descriptionInput, { target: { value: "Updated task" } });
    // Submit update
    expect(descriptionInput.value).toBe("Updated task");
    fireEvent.click(screen.getByRole("button", { name: /Update Task/i }));
    // Assert input value updated and patch API called with correct payload
    await waitFor(() => {
      expect(client.api.tasks.patch).toHaveBeenCalledWith(
        1,
        expect.objectContaining({ description: "Updated task" })
      );
    });
  });

  it("marks a task as completed", async () => {
    renderTasks();
    // Wait for edit buttons to appear
    await waitFor(() => {
      expect(screen.getAllByTitle("Edit task")[1]).toBeInTheDocument();
    });
    // Open edit modal for second task (in-progress)
    fireEvent.click(screen.getAllByTitle("Edit task")[1]);
    // Change status to Done
    const statusSelect = screen.getByLabelText("Status") as HTMLSelectElement;
    fireEvent.change(statusSelect, { target: { value: "done" } });
    // Submit update
    fireEvent.click(screen.getByRole("button", { name: /Update Task/i }));
    // Assert patch API called with correct args
    await waitFor(() => {
      expect(client.api.tasks.patch).toHaveBeenCalledWith(
        2,
        expect.objectContaining({ status: "done" })
      );
    });
  });

  it("matches Tasks component snapshot", async () => {
    const { container } = renderTasks();
    await waitFor(() => {
      expect(screen.getByText("Tasks")).toBeInTheDocument();
    });
    expect(container).toMatchSnapshot();
  });
});
