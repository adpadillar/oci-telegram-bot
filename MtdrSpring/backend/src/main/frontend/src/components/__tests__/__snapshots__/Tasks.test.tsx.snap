// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tasks Component > matches Tasks component snapshot 1`] = `
<div>
  <div
    class="p-4 sm:p-6 bg-gray-50 min-h-screen"
  >
    <div
      class="bg-white rounded-xl shadow-sm border border-gray-200 p-4 sm:p-6 mb-6"
    >
      <div
        class="flex flex-col sm:flex-row justify-between items-start sm:items-center gap-4 mb-6"
      >
        <h1
          class="text-xl sm:text-2xl font-bold text-gray-800 flex items-center gap-2"
        >
          <svg
            class="lucide lucide-tag text-blue-500 h-5 w-5 sm:h-6 sm:w-6"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12.586 2.586A2 2 0 0 0 11.172 2H4a2 2 0 0 0-2 2v7.172a2 2 0 0 0 .586 1.414l8.704 8.704a2.426 2.426 0 0 0 3.42 0l6.58-6.58a2.426 2.426 0 0 0 0-3.42z"
            />
            <circle
              cx="7.5"
              cy="7.5"
              fill="currentColor"
              r=".5"
            />
          </svg>
          Tasks
          <span
            class="text-sm font-normal text-gray-500 ml-2"
          >
            2
             
            tasks
          </span>
        </h1>
        <div
          class="flex flex-col sm:flex-row items-stretch sm:items-center gap-3 w-full sm:w-auto"
        >
          <button
            class="bg-blue-500 text-white px-3 py-2 rounded-md flex items-center justify-center gap-2 hover:bg-blue-600 transition-colors whitespace-nowrap"
          >
            <svg
              class="lucide lucide-download"
              fill="none"
              height="16"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M21 15v4a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2v-4"
              />
              <polyline
                points="7 10 12 15 17 10"
              />
              <line
                x1="12"
                x2="12"
                y1="15"
                y2="3"
              />
            </svg>
            <span
              class="hidden sm:inline"
            >
              Download Task List
            </span>
            <span
              class="sm:hidden"
            >
              Descargar
            </span>
          </button>
          <div
            class="relative w-full sm:w-64"
          >
            <div
              class="absolute inset-y-0 left-0 pl-3 flex items-center pointer-events-none"
            >
              <svg
                class="lucide lucide-search h-4 w-4 text-gray-400"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  cx="11"
                  cy="11"
                  r="8"
                />
                <path
                  d="m21 21-4.3-4.3"
                />
              </svg>
            </div>
            <input
              class="pl-10 pr-4 py-2 w-full border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-transparent"
              placeholder="Search tasks..."
              type="text"
              value=""
            />
          </div>
          <div
            class="flex items-center gap-2"
          >
            <div
              class="bg-gray-100 p-1 rounded-md flex"
            >
              <button
                class="p-2 rounded-md flex items-center transition-colors text-gray-500 hover:text-gray-700"
                title="Kanban view"
              >
                <svg
                  class="lucide lucide-layout-grid"
                  fill="none"
                  height="18"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="18"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <rect
                    height="7"
                    rx="1"
                    width="7"
                    x="3"
                    y="3"
                  />
                  <rect
                    height="7"
                    rx="1"
                    width="7"
                    x="14"
                    y="3"
                  />
                  <rect
                    height="7"
                    rx="1"
                    width="7"
                    x="14"
                    y="14"
                  />
                  <rect
                    height="7"
                    rx="1"
                    width="7"
                    x="3"
                    y="14"
                  />
                </svg>
              </button>
              <button
                class="p-2 rounded-md flex items-center transition-colors bg-white shadow-sm text-blue-500"
                title="Table view"
              >
                <svg
                  class="lucide lucide-list"
                  fill="none"
                  height="18"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="18"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M3 12h.01"
                  />
                  <path
                    d="M3 18h.01"
                  />
                  <path
                    d="M3 6h.01"
                  />
                  <path
                    d="M8 12h13"
                  />
                  <path
                    d="M8 18h13"
                  />
                  <path
                    d="M8 6h13"
                  />
                </svg>
              </button>
            </div>
            <div
              class="relative"
            >
              <button
                class="px-3 py-2 rounded-md flex items-center gap-1 transition-colors text-gray-600 hover:bg-gray-100"
              >
                <svg
                  class="lucide lucide-filter"
                  fill="none"
                  height="16"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <polygon
                    points="22 3 2 3 10 12.46 10 19 14 21 14 12.46 22 3"
                  />
                </svg>
                <span
                  class="hidden sm:inline"
                >
                  Filter
                </span>
              </button>
            </div>
            <button
              class="bg-blue-500 text-white px-3 py-2 rounded-md flex items-center justify-center gap-2 hover:bg-blue-600 transition-colors whitespace-nowrap"
            >
              <svg
                class="lucide lucide-plus"
                fill="none"
                height="16"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5 12h14"
                />
                <path
                  d="M12 5v14"
                />
              </svg>
              <span
                class="hidden sm:inline"
              >
                Add Task
              </span>
              <span
                class="sm:hidden"
              >
                Add
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        class="mt-4 overflow-x-auto rounded-lg border border-gray-200"
      >
        <table
          class="min-w-full divide-y divide-gray-200 bg-white"
        >
          <thead
            class="bg-gray-50"
          >
            <tr>
              <th
                class="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider"
                scope="col"
              >
                Name
              </th>
              <th
                class="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider"
                scope="col"
              >
                Category
              </th>
              <th
                class="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider"
                scope="col"
              >
                Sprint
              </th>
              <th
                class="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider"
                scope="col"
              >
                Due Date
              </th>
              <th
                class="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider"
                scope="col"
              >
                Estimate / Real
              </th>
              <th
                class="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider"
                scope="col"
              >
                Assigned To
              </th>
              <th
                class="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider"
                scope="col"
              >
                Status
              </th>
              <th
                class="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider"
                scope="col"
              >
                Actions
              </th>
            </tr>
          </thead>
          <tbody
            class="bg-white divide-y divide-gray-200"
          >
            <tr
              class="hover:bg-gray-50"
            >
              <td
                class="px-6 max-w-[40vw] md:max-w-[32rem] truncate py-4 whitespace-nowrap text-sm font-medium text-gray-900"
              >
                Fix login bug
              </td>
              <td
                class="px-6 py-4 whitespace-nowrap"
              >
                <span
                  class="px-2 py-1 inline-flex items-center text-xs leading-5 font-semibold rounded-full 
                      bg-red-100 text-red-800"
                >
                  <svg
                    class="lucide lucide-bug mr-1"
                    fill="none"
                    height="12"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="12"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="m8 2 1.88 1.88"
                    />
                    <path
                      d="M14.12 3.88 16 2"
                    />
                    <path
                      d="M9 7.13v-1a3.003 3.003 0 1 1 6 0v1"
                    />
                    <path
                      d="M12 20c-3.3 0-6-2.7-6-6v-3a4 4 0 0 1 4-4h4a4 4 0 0 1 4 4v3c0 3.3-2.7 6-6 6"
                    />
                    <path
                      d="M12 20v-9"
                    />
                    <path
                      d="M6.53 9C4.6 8.8 3 7.1 3 5"
                    />
                    <path
                      d="M6 13H2"
                    />
                    <path
                      d="M3 21c0-2.1 1.7-3.9 3.8-4"
                    />
                    <path
                      d="M20.97 5c0 2.1-1.6 3.8-3.5 4"
                    />
                    <path
                      d="M22 13h-4"
                    />
                    <path
                      d="M17.2 17c2.1.1 3.8 1.9 3.8 4"
                    />
                  </svg>
                  Bug
                </span>
              </td>
              <td
                class="px-6 py-4 whitespace-nowrap text-sm text-gray-500"
              >
                <span
                  class="bg-blue-50 text-blue-600 px-2 py-0.5 rounded-full text-xs"
                >
                  Sprint 1
                </span>
              </td>
              <td
                class="px-6 py-4 whitespace-nowrap text-sm"
              >
                <button
                  class="text-gray-400 hover:text-gray-600 flex items-center text-xs border border-dashed border-gray-300 rounded-full px-2 py-1 transition-colors"
                  title="Set due date"
                >
                  <svg
                    class="lucide lucide-calendar-check mr-1"
                    fill="none"
                    height="12"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="12"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M8 2v4"
                    />
                    <path
                      d="M16 2v4"
                    />
                    <rect
                      height="18"
                      rx="2"
                      width="18"
                      x="3"
                      y="4"
                    />
                    <path
                      d="M3 10h18"
                    />
                    <path
                      d="m9 16 2 2 4-4"
                    />
                  </svg>
                  <span>
                    Set date
                  </span>
                </button>
              </td>
              <td
                class="px-6 py-4 whitespace-nowrap text-sm"
              >
                <div
                  class="flex items-center gap-1"
                >
                  <svg
                    class="lucide lucide-clock text-gray-400"
                    fill="none"
                    height="14"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="14"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <circle
                      cx="12"
                      cy="12"
                      r="10"
                    />
                    <polyline
                      points="12 6 12 12 16 14"
                    />
                  </svg>
                  <span
                    class="font-medium text-gray-700"
                  >
                    4
                    h
                  </span>
                  <span
                    class="text-gray-400 mx-1"
                  >
                    /
                  </span>
                  <span
                    class="font-medium text-red-600"
                  >
                    5
                    h
                  </span>
                </div>
              </td>
              <td
                class="px-6 py-4 whitespace-nowrap text-sm text-gray-500"
              >
                <div
                  class="flex items-center"
                >
                  <img
                    alt="1's avatar"
                    class="w-6 h-6 rounded-full mr-2 border border-gray-200"
                    src="https://api.dicebear.com/7.x/avataaars/svg?seed=1"
                  />
                  <span>
                    Ana
                  </span>
                </div>
              </td>
              <td
                class="px-6 py-4 whitespace-nowrap"
              >
                <span
                  class="px-2 py-1 inline-flex items-center gap-1 text-xs leading-5 font-semibold rounded-full bg-green-100 text-green-800"
                >
                  Done
                </span>
              </td>
              <td
                class="px-6 py-4 flex space-x-2 whitespace-nowrap text-sm text-gray-500"
              >
                <button
                  class="text-blue-600 hover:text-blue-800 transition-colors p-1 rounded-md hover:bg-blue-50"
                  title="Edit task"
                >
                  <svg
                    class="lucide lucide-pencil"
                    fill="none"
                    height="16"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M21.174 6.812a1 1 0 0 0-3.986-3.987L3.842 16.174a2 2 0 0 0-.5.83l-1.321 4.352a.5.5 0 0 0 .623.622l4.353-1.32a2 2 0 0 0 .83-.497z"
                    />
                    <path
                      d="m15 5 4 4"
                    />
                  </svg>
                </button>
                <button
                  class="text-red-600 hover:text-red-800 p-1 rounded-md hover:bg-red-50"
                  title="Delete task"
                >
                  <svg
                    class="lucide lucide-trash2"
                    fill="none"
                    height="16"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M3 6h18"
                    />
                    <path
                      d="M19 6v14c0 1-1 2-2 2H7c-1 0-2-1-2-2V6"
                    />
                    <path
                      d="M8 6V4c0-1 1-2 2-2h4c1 0 2 1 2 2v2"
                    />
                    <line
                      x1="10"
                      x2="10"
                      y1="11"
                      y2="17"
                    />
                    <line
                      x1="14"
                      x2="14"
                      y1="11"
                      y2="17"
                    />
                  </svg>
                </button>
              </td>
            </tr>
            <tr
              class="hover:bg-gray-50"
            >
              <td
                class="px-6 max-w-[40vw] md:max-w-[32rem] truncate py-4 whitespace-nowrap text-sm font-medium text-gray-900"
              >
                Implement UI
              </td>
              <td
                class="px-6 py-4 whitespace-nowrap"
              >
                <span
                  class="px-2 py-1 inline-flex items-center text-xs leading-5 font-semibold rounded-full 
                      bg-purple-100 text-purple-800"
                >
                  <svg
                    class="lucide lucide-sparkles mr-1"
                    fill="none"
                    height="12"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="12"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M9.937 15.5A2 2 0 0 0 8.5 14.063l-6.135-1.582a.5.5 0 0 1 0-.962L8.5 9.936A2 2 0 0 0 9.937 8.5l1.582-6.135a.5.5 0 0 1 .963 0L14.063 8.5A2 2 0 0 0 15.5 9.937l6.135 1.581a.5.5 0 0 1 0 .964L15.5 14.063a2 2 0 0 0-1.437 1.437l-1.582 6.135a.5.5 0 0 1-.963 0z"
                    />
                    <path
                      d="M20 3v4"
                    />
                    <path
                      d="M22 5h-4"
                    />
                    <path
                      d="M4 17v2"
                    />
                    <path
                      d="M5 18H3"
                    />
                  </svg>
                  Feature
                </span>
              </td>
              <td
                class="px-6 py-4 whitespace-nowrap text-sm text-gray-500"
              >
                <span
                  class="bg-blue-50 text-blue-600 px-2 py-0.5 rounded-full text-xs"
                >
                  Sprint 1
                </span>
              </td>
              <td
                class="px-6 py-4 whitespace-nowrap text-sm"
              >
                <button
                  class="text-gray-400 hover:text-gray-600 flex items-center text-xs border border-dashed border-gray-300 rounded-full px-2 py-1 transition-colors"
                  title="Set due date"
                >
                  <svg
                    class="lucide lucide-calendar-check mr-1"
                    fill="none"
                    height="12"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="12"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M8 2v4"
                    />
                    <path
                      d="M16 2v4"
                    />
                    <rect
                      height="18"
                      rx="2"
                      width="18"
                      x="3"
                      y="4"
                    />
                    <path
                      d="M3 10h18"
                    />
                    <path
                      d="m9 16 2 2 4-4"
                    />
                  </svg>
                  <span>
                    Set date
                  </span>
                </button>
              </td>
              <td
                class="px-6 py-4 whitespace-nowrap text-sm"
              >
                <div
                  class="flex items-center gap-1"
                >
                  <svg
                    class="lucide lucide-clock text-gray-400"
                    fill="none"
                    height="14"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="14"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <circle
                      cx="12"
                      cy="12"
                      r="10"
                    />
                    <polyline
                      points="12 6 12 12 16 14"
                    />
                  </svg>
                  <span
                    class="font-medium text-gray-700"
                  >
                    6
                    h
                  </span>
                  <span
                    class="text-gray-400 mx-1"
                  >
                    /
                  </span>
                  <span
                    class="font-medium text-green-600"
                  >
                    -
                    h
                  </span>
                </div>
              </td>
              <td
                class="px-6 py-4 whitespace-nowrap text-sm text-gray-500"
              >
                <div
                  class="flex items-center"
                >
                  <img
                    alt="2's avatar"
                    class="w-6 h-6 rounded-full mr-2 border border-gray-200"
                    src="https://api.dicebear.com/7.x/avataaars/svg?seed=2"
                  />
                  <span>
                    Luis
                  </span>
                </div>
              </td>
              <td
                class="px-6 py-4 whitespace-nowrap"
              >
                <span
                  class="px-2 py-1 inline-flex items-center gap-1 text-xs leading-5 font-semibold rounded-full bg-yellow-100 text-yellow-800"
                >
                  In Progress
                </span>
              </td>
              <td
                class="px-6 py-4 flex space-x-2 whitespace-nowrap text-sm text-gray-500"
              >
                <button
                  class="text-blue-600 hover:text-blue-800 transition-colors p-1 rounded-md hover:bg-blue-50"
                  title="Edit task"
                >
                  <svg
                    class="lucide lucide-pencil"
                    fill="none"
                    height="16"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M21.174 6.812a1 1 0 0 0-3.986-3.987L3.842 16.174a2 2 0 0 0-.5.83l-1.321 4.352a.5.5 0 0 0 .623.622l4.353-1.32a2 2 0 0 0 .83-.497z"
                    />
                    <path
                      d="m15 5 4 4"
                    />
                  </svg>
                </button>
                <button
                  class="text-red-600 hover:text-red-800 p-1 rounded-md hover:bg-red-50"
                  title="Delete task"
                >
                  <svg
                    class="lucide lucide-trash2"
                    fill="none"
                    height="16"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M3 6h18"
                    />
                    <path
                      d="M19 6v14c0 1-1 2-2 2H7c-1 0-2-1-2-2V6"
                    />
                    <path
                      d="M8 6V4c0-1 1-2 2-2h4c1 0 2 1 2 2v2"
                    />
                    <line
                      x1="10"
                      x2="10"
                      y1="11"
                      y2="17"
                    />
                    <line
                      x1="14"
                      x2="14"
                      y1="11"
                      y2="17"
                    />
                  </svg>
                </button>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
